src/boot_O2_g3/fault.c

```diff
@@ -572,354 +696,440 @@ void Fault_DrawMemDump(u32 pc, u32 sp, u32 unk0, u32 unk1) {
-    osStartThread(t);
+void Fault_ResumeThread(OSThread* thread) {
+    thread->context.cause = 0;
+    thread->context.fpcsr = 0;
+    thread->context.pc += sizeof(u32);
+    *(u32*)thread->context.pc = 0x0000000D; // write in a break instruction
+    osWritebackDCache((void*)thread->context.pc, 4);
+    osInvalICache((void*)thread->context.pc, 4);
+    osStartThread(thread);
 }
 
-void Fault_CommitFB(void) {
-    u16* fb;
+void Fault_DisplayFrameBuffer(void) {
+    void* fb;
 
     osViSetYScale(1.0f);
     osViSetMode(&osViModeNtscLan1);
-    osViSetSpecialFeatures(0x42); // gama_disable|dither_fliter_enable_aa_mode3_disable
+    osViSetSpecialFeatures(OS_VI_GAMMA_OFF | OS_VI_DITHER_FILTER_ON);
     osViBlack(false);
 
-    if (sFaultContext->fb) {
-        fb = sFaultContext->fb;
+    if (sFaultInstance->fb) {
+        fb = sFaultInstance->fb;
     } else {
-        fb = (u16*)osViGetNextFramebuffer();
-        if ((u32)fb == 0x80000000) {
-            fb = (u16*)((osMemSize | 0x80000000) - 0x25800);
+        fb = osViGetNextFramebuffer();
+        if ((uintptr_t)fb == K0BASE) {
+            fb = (void*)(PHYS_TO_K0(osMemSize) - SCREEN_HEIGHT * SCREEN_WIDTH * sizeof(u16));
         }
     }
 
     osViSwapBuffer(fb);
     FaultDrawer_SetDrawerFB(fb, SCREEN_WIDTH, SCREEN_HEIGHT);
 }
 
+/**
+ * Runs all registered fault clients. Each fault client displays a page
+ * on the crash screen.
+ */
 void Fault_ProcessClients(void) {
-    FaultClient* iter = sFaultContext->clients;
+    FaultClient* client = sFaultInstance->clients;
     s32 idx = 0;
 
-    while (iter != NULL) {
-        if (iter->callback) {
+    while (client != NULL) {
+        if (client->callback != NULL) {
             Fault_FillScreenBlack();
             FaultDrawer_SetCharPad(-2, 0);
-            FaultDrawer_Printf("\x1A\x38"
-                               "CallBack (%d) %08x %08x %08x\n"
-                               "\x1A\x37",
-                               idx++, iter, iter->param0, iter->param1);
+            FaultDrawer_Printf(FAULT_COLOR(DARK_GRAY) "CallBack (%d) %08x %08x %08x\n" FAULT_COLOR(WHITE), idx++,
+                               client, client->arg0, client->arg1);
             FaultDrawer_SetCharPad(0, 0);
-            iter->callback(iter->param0, iter->param1);
+            client->callback(client->arg0, client->arg1);
             Fault_WaitForInput();
-            Fault_CommitFB();
+            Fault_DisplayFrameBuffer();
         }
-        iter = iter->next;
+        client = client->next;
     }
 }
 
 #ifdef NON_MATCHING
 // needs in-function static bss
 void Fault_SetOptionsFromController3(void) {
     static u32 faultCustomOptions;
-    Input* input3 = &sFaultContext->padInput[3];
+    Input* input3 = &sFaultInstance->inputs[3];
     u32 pad;
-    u32 graphPC;
-    u32 graphRA;
-    u32 graphSP;
+    uintptr_t pc;
+    uintptr_t ra;
+    uintptr_t sp;
 
     if (CHECK_BTN_ALL(input3->press.button, 0x80)) {
```

Comment:
https://discord.com/channels/688807550715560050/688851976746041391/1084273960721592360
Could probably use a comment that this is a neutral reset.